### PR TITLE
LA-484 Make offline mode for Shared Space Document in SharedSpace

### PR DESCRIPTION
**Pull request by dab246**
https://user-images.githubusercontent.com/80730648/122379062-b68f0a80-cf90-11eb-97fa-f69ee7afc106.mp4

### CODE REVIEW

File: domain/lib/src/model/sharedspacedocument/work_group_folder.dart [hunk 1]
```diff
@@ -33,6 +33,9 @@
 import 'package:domain/domain.dart';
 
 class WorkGroupFolder extends WorkGroupNode {
+
+  final List<TreeNode> listTreeNode;
```
Comment (hoangdat) [inline domain/lib/src/model/sharedspacedocument/work_group_folder.dart:37]: should be define in `WorkGroupNode`

Comment (hoangdat) [reply domain/lib/src/model/sharedspacedocument/work_group_folder.dart:37]: name is `treePath` to easier in parsing.

File: domain/lib/src/usecases/shared_space_document/get_work_group_node_detail_interactor.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+
+import 'package:dartz/dartz.dart';
+import 'package:domain/domain.dart';
+import 'package:domain/src/state/failure.dart';
+import 'package:domain/src/state/success.dart';
+import 'package:domain/src/usecases/shared_space_document/get_work_group_node_detail_view_state.dart';
+
+class GetWorkGroupNodeDetailInteractor {
```
Comment (hoangdat) [inline domain/lib/src/usecases/shared_space_document/get_work_group_node_detail_interactor.dart:38]: don't need to add new more

File: domain/lib/src/model/sharedspacedocument/work_group_folder.dart (hunk 1 above)
Comment (dab246) [reply domain/lib/src/model/sharedspacedocument/work_group_folder.dart:37]: ok

File: domain/lib/src/usecases/shared_space_document/get_work_group_node_detail_interactor.dart (hunk 2 above)
Comment (dab246) [reply domain/lib/src/usecases/shared_space_document/get_work_group_node_detail_interactor.dart:38]: ok

File: data/lib/src/extensions/tree_node_extension.dart [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+
+import 'package:data/src/local/model/work_group_node_cache.dart';
+import 'package:domain/domain.dart';
+
+extension TreeNodeExtension on TreeNode {
+  WorkGroupNodeCache toWorkGroupNodeCache(SharedSpaceId sharedSpaceId, WorkGroupNodeId parentId) {
+    return WorkGroupNodeCache(
```
Comment (hoangdat) [inline data/lib/src/extensions/tree_node_extension.dart:38]: should have the type of the node: folder or document?

File: data/lib/src/local/database_client.dart [hunk 4]
```diff
@@ -108,5 +124,25 @@ class DatabaseClient {
...
+  Future<List<Map<String, dynamic>>> getListDataWithCondition(String tableName, String condition, List<dynamic> values) async {
+    final db = await database;
+    return await db.query(
+        tableName,
+        where: condition,
+        whereArgs: values);
+  }
+
+  Future insertMultipleData(String tableName, List<Map<String, dynamic>> mapObjects) async {
+    final bat = await batch;
```
Comment (hoangdat) [inline data/lib/src/local/database_client.dart:131]: `batchInsert`

File: domain/lib/src/usecases/get_child_nodes/get_all_child_nodes_interactor.dart [hunk 5]
```diff
@@ -48,14 +48,28 @@ class GetAllChildNodesInteractor {
...
+
+      if (newChildNodes.every((element) => element.type == WorkGroupNodeType.DOCUMENT_REVISION)) {
+        newChildNodes.sort((node1, node2) {
+          return node2.creationDate.compareTo(node1.creationDate);
         });
       } else {
         childNodes.sort((node1, node2) {
           return node2!.modificationDate.compareTo(node1!.modificationDate);
+        newChildNodes.sort((node1, node2) {
+          return node2.modificationDate.compareTo(node1.modificationDate);
```
[the remark this answers is not in the document]
Comment (dab246) [reply domain/lib/src/usecases/get_child_nodes/get_all_child_nodes_interactor.dart]: I did not write this. I just fixed it via null-safety

File: domain/lib/src/model/sharedspacedocument/work_group_document.dart [hunk 6]
```diff
@@ -40,6 +40,8 @@ class WorkGroupDocument extends WorkGroupNode {
   final DateTime uploadDate;
   final bool hasRevision;
   final String sha256sum;
+  final String localPath;
```
[the remark this answers is not in the document]
Comment (dab246) [reply domain/lib/src/model/sharedspacedocument/work_group_document.dart:43]: It is a nullable. I fixed it in the last commits.

File: data/lib/src/datasource/quota_datasource.dart [hunk 7]
```diff
@@ -32,5 +32,5 @@
 import 'package:domain/domain.dart';
 
 abstract class QuotaDataSource {
-  Future<AccountQuota> findQuota(QuotaId quotaUuid);
+  Future<AccountQuota> findQuota(QuotaId? quotaUuid);
```
Comment (hoangdat) [inline data/lib/src/datasource/quota_datasource.dart:35]: Must have `quotaId` to find quota.

File: data/lib/src/datasource/shared_space_document_datasource.dart [hunk 8]
```diff
@@ -81,8 +81,8 @@ abstract class SharedSpaceDocumentDataSource {
     RenameWorkGroupNodeRequest renameRequest
   );
 
-  Future<WorkGroupNode> getWorkGroupNode(
-    SharedSpaceId sharedSpaceId,
+  Future<WorkGroupNode?> getWorkGroupNode(
+    SharedSpaceId? sharedSpaceId,
```
Comment (hoangdat) [inline data/lib/src/datasource/shared_space_document_datasource.dart:85]: non-null

File: data/lib/src/datasource/shared_space_document_datasource.dart [hunk 9]
```diff
@@ -95,10 +95,12 @@ abstract class SharedSpaceDocumentDataSource {
     Uri baseUrl);
 
   Future<bool> makeAvailableOfflineSharedSpaceDocument(
-    SharedSpaceNodeNested sharedSpaceNodeNested,
+    SharedSpaceNodeNested? sharedSpaceNodeNested,
```
Comment (hoangdat) [inline data/lib/src/datasource/shared_space_document_datasource.dart:98]: why not only `sharedSpaceId` only?

File: data/lib/src/extensions/shared_space_node_nested_extension.dart [hunk 10]
```diff
@@ -36,15 +36,15 @@ extension SharedSpaceNodeNestedExtension on SharedSpaceNodeNested {
   SharedSpaceCache toSharedSpaceDto() {
     return SharedSpaceCache(
         sharedSpaceId,
-        sharedSpaceRole != null ? sharedSpaceRole.sharedSpaceRoleId : null,
-        sharedSpaceRole != null ? sharedSpaceRole.name : null,
-        sharedSpaceRole != null ? sharedSpaceRole.enabled : null,
+        sharedSpaceRole != null ? sharedSpaceRole!.sharedSpaceRoleId : null,
```
Comment (hoangdat) [inline data/lib/src/extensions/shared_space_node_nested_extension.dart:39]: why `role` can be null?

File: data/lib/src/extensions/tree_node_extension.dart [hunk 11]
```diff
@@ -33,11 +33,11 @@ import 'package:data/src/local/model/work_group_node_cache.dart';
 import 'package:domain/domain.dart';
 
 extension TreeNodeExtension on TreeNode {
-  WorkGroupNodeCache toWorkGroupNodeCache(SharedSpaceId sharedSpaceId, WorkGroupNodeId parentId) {
+  WorkGroupNodeCache toWorkGroupNodeCache(SharedSpaceId sharedSpaceId, WorkGroupNodeId? parentId) {
     return WorkGroupNodeCache(
         workGroupNodeId,
         sharedSpaceId,
-        parentId,
+        parentId ?? WorkGroupNodeId(''),
```
Comment (hoangdat) [inline data/lib/src/extensions/tree_node_extension.dart:41]: I prefer null or a constant. 
```
const WorkGroupNodeId parent = WorkGroupNodeId('');
```

File: data/lib/src/extensions/work_group_document_extension.dart [hunk 12]
```diff
@@ -76,8 +76,8 @@ extension WorkGroupDocumentExtension on WorkGroupDocument {
         uploadDate,
         hasRevision,
         sha256sum,
-        localPath: localPath,
-        syncOfflineState: syncOfflineState
+        localPath: localPath ?? '',
```
Comment (hoangdat) [inline data/lib/src/extensions/work_group_document_extension.dart:80]: `localPath` should be nullable

File: data/lib/src/datasource/shared_space_document_datasource.dart (hunk 9 above)
Comment (dab246) [reply data/lib/src/datasource/shared_space_document_datasource.dart:98]: We need to save this SharedSpace's information in the database too

**Review by huynguyennovem — APPROVED** [review]
/lgtm

**Pull request merged**